src/modules-by-example.adoc

```diff
@@ -0,0 +1,320 @@
+[[sec:moduleexample]]
+== Modules by Example
+
+:nrm: subs="+normal"
+
+Within the spectrum of Ion 1.1's new encoding facilities, macros lie at the bottom, affecting the
+serialized byte-stream.  The prior chapter explored macro definitions while ignoring the contexts
+within which those definitions exist.  This chapter covers that context top-down.
+
+
+=== Encoding Environment 1.0
+
+An Ion _document_ is a stream of octets conforming to either the Ion text or binary specification.
+(For our purposes here, a document does not necessarily exist as a file, and isn’t necessarily
+finite.)  The interpretation of those octets is guided by an _encoding environment_, the
+context-specific data maintained by an Ion implementation while encoding or decoding a document.
+The Ion 1.0 encoding environment is effectively just the local symbol table.
+
+The encoding environment is controlled by _directives_ embedded in the document at top-level.
+These directives, also known as “system values”, are encoding artifacts and are not part of the
+application data model.
+
+Ion 1.0 has two forms of directives:
+
+* An Ion Version Marker (IVM) resets the environment to the default provided by that version of
+Ion.
+* An `$ion_symbol_table` struct defines a new environment that takes effect immediately after the
+struct closes.
+
+A _segment_ is a contiguous portion of a document that uses the same encoding environment.
+Segment boundaries are caused by directives: an IVM starts a new segment, while an
+`$ion_symbol_table` struct ends a segment, defining a new one that starting immediately
+afterwards.
+
+TODO Ion text docs always start with a 1.0 segment until an IVM is encountered.
+
+
+=== Modules from the Outside
+
+In Ion 1.1, you define, share, and install symbols and macros using _encoding modules_.
+The logical interface to a module has three main components: a spec version, a symbol
+table, and a macro table.
+
+A module's _spec version_ indicates which Ion specification it uses. This ensures the module has
+stable semantics over time.  Spec versions are expressed as Ion version markers, and a module can
+only be used in segments encoded with that version or later.
+
+// TODO The above may be too strict; use solely for symbols could be more relaxed.
+
+A module's _exported symbol table_ is simply a sequence of strings. These denote the
+text of symbols, and are equivalent in meaning to the `symbols` list of an Ion 1.0 shared symbol
+table.
+
+A module's _exported macro table_ is a sequence of <name, macro> pairs.  Names can be
+null, in which case the corresponding macro can be referenced by its zero-based index in the
+table, known as its _exported address_.  Non-null names in the table must all be unique, so that
+a name-to-macro mapping function is well-defined.
+
+TIP: Macros have their own "object identity" independent of the names that map to them.  It's
+possible for the same macro to have multiple addresses and/or names.
```

Comment:
I think its more than an implementation detail, since the rule around reference ambiguity requires identity of macros.  Specifically, an alias to a macro has the same identity, as does an re-export from an importing module.  Copying the macro—or otherwise assigning a second identity with the same definition—would introduce ambiguity.

Reply:
In #214 I change this to simply read "their own identity".